File: src/mvc/cpp/reordable_treeview/tree_model.cpp
#include "tree_model.h"
#include <QStringList>
#include <exception>

QVector<QVariant> stringListToVariantList(const QStringList &data) {
  QVector<QVariant> variantData;
  variantData.reserve(data.count());
  for (auto &string : data) {
    variantData.append(QVariant(string));
  }
  return variantData;
}

TreeItem::TreeItem(const QVector<QVariant> &data, TreeItem *parent)
    : itemData(data), parentItem(parent) {}

TreeItem::TreeItem(const QStringList &data, TreeItem *parent)
    : TreeItem(stringListToVariantList(data), parent) {}

TreeItem::~TreeItem() { qDeleteAll(childItems); }

void TreeItem::appendChild(TreeItem *item) {
  if (!item) {
    return;
  }

  auto currentParent = item->parent();
  if (currentParent) {
    currentParent->removeChild(item);
  }
  item->setParent(this);
  childItems.append(item);
}

TreeItem *TreeItem::child(int row) {
  if (row < 0 || row >= childItems.size()) {
    return nullptr;
  }
  return childItems.at(row);
}

int TreeItem::childFromRow(TreeItem *child) {
  return childItems.indexOf(child);
}

QVector<TreeItem *> TreeItem::childrenItems() { return childItems; }

int TreeItem::rowCount() const { return childItems.count(); }

int TreeItem::columnCount() const { return itemData.count(); }

int TreeItem::childrenColumnCount() const {
  int maxCount = 0;
  for (auto &child : childItems) {
    maxCount = std::max(maxCount, child->columnCount());
  }
  return maxCount;
}

QVariant TreeItem::data(int column) const {

  if (column < 0 || column >= columnCount()) {
    return QVariant();
  }
  return itemData.at(column);
}

QVector<QVariant> TreeItem::data() const { return itemData; }

TreeItem *TreeItem::parent() { return parentItem; }

int TreeItem::row() const {
  if (parentItem) {
    return parentItem->childItems.indexOf(const_cast<TreeItem *>(this));
  }

  return 0;
}

void TreeItem::removeRow(int row) {
  if (row < 0 || row >= rowCount()) {
    return;
  }
  childItems.remove(row);
}

void TreeItem::removeRows(int row, int count) {
  for (int i = 0; i < count; ++i) {
    childItems.remove(row);
  }
}

void TreeItem::removeChild(TreeItem *child) { childItems.removeOne(child); }

void TreeItem::removeAllChildren() { childItems.clear(); }

void TreeItem::setData(const QVector<QVariant> &data) { itemData = data; }

void TreeItem::setParent(TreeItem *parent) {
  if (!parent) {
    return;
  }

  parentItem = parent;
}

void TreeItem::removeColumns(int position, int columns) {
  if (position < 0 || position + columns > itemData.size()) {
    throw std::out_of_range("removeColumns: invalid arguments");
  }

  for (int column = 0; column < columns; column++) {
    itemData.remove(position);
  }

  for (const auto &child : childItems) {
    child->removeColumns(position, columns);
  }
}

void TreeItem::insertColumns(int position, int columns) {
  if (position < 0 || position > itemData.size()) {
    throw std::out_of_range("insertColumns: invalid arguments");
  }

  for (int column = 0; column < columns; ++column) {
    itemData.insert(position, QVariant());
  }

  for (const auto &child : childItems) {
    child->insertColumns(position, columns);
  }
}

void TreeItem::insertRows(int position, int count, int columns) {
  if (position < 0 || position > childItems.size()) {
    throw std::out_of_range("insertRows: invalid arguments");
  }

  for (int row = 0; row < count; ++row) {
    QVector<QVariant> data(columns);
    auto item = new TreeItem(data, this);
    childItems.insert(position, item);
  }
}

void TreeItem::setData(int column, const QVariant &value) {
  if (column < 0 || column >= itemData.size()) {
    throw std::out_of_range("setData: invalid arguments");
  }

  itemData[column] = value;
}

TreeModel::TreeModel(const QStringList &headers, QObject *parent)
    : QAbstractItemModel(parent), readOnly(false) {
  QVector<QVariant> rootData;
  for (const QString &header : headers) {
    rootData << header;
  }

  rootItem = new TreeItem(rootData);
}

TreeModel::~TreeModel() { delete rootItem; }

void TreeModel::setReadOnly(bool flag) { readOnly = flag; }

bool TreeModel::isReadOnly() const { return readOnly; }

int TreeModel::columnCount(const QModelIndex &parent) const {
  if (parent.isValid()) {
    return static_cast<TreeItem *>(parent.internalPointer())
        ->childrenColumnCount();
  }
  return rootItem->columnCount();
}

QVariant TreeModel::data(const QModelIndex &index, int role) const {
  if (!index.isValid()) {
    return QVariant();
  }

  if (role == Qt::DisplayRole || role == Qt::EditRole) {
    auto item = static_cast<TreeItem *>(index.internalPointer());
    return item->data(index.column());
  }

  return QVariant();
}

Qt::ItemFlags TreeModel::flags(const QModelIndex &index) const {

  if (!index.isValid()) {
    return Qt::NoItemFlags;
  }

  if (readOnly) {
    return QAbstractItemModel::flags(index);
  }

  if (hasChildren(index)) {
    return Qt::ItemIsEnabled | Qt::ItemIsSelectable;
  } else {
    return Qt::ItemIsEnabled | Qt::ItemIsSelectable | Qt::ItemIsEditable;
  }
  return Qt::ItemIsEnabled;
}

QVariant TreeModel::headerData(int section, Qt::Orientation orientation,
                               int role) const {
  if (orientation == Qt::Horizontal && role == Qt::DisplayRole) {
    return rootItem->data(section);
  }

  return QVariant();
}

QModelIndex TreeModel::index(int row, int column,
                             const QModelIndex &parent) const {
  if (!hasIndex(row, column, parent)) {
    return QModelIndex();
  }

  TreeItem *parentItem;

  if (!parent.isValid()) {
    parentItem = rootItem;
  } else {
    parentItem = static_cast<TreeItem *>(parent.internalPointer());
  }

  auto childItem = parentItem->child(row);

  if (childItem) {
    return createIndex(row, column, childItem);
  }

  return QModelIndex();
}

QModelIndex TreeModel::parent(const QModelIndex &index) const {
  if (!index.isValid()) {
    return QModelIndex();
  }

  auto childItem = static_cast<TreeItem *>(index.internalPointer());
  auto parentItem = childItem->parent();

  if (parentItem == rootItem) {
    return QModelIndex();
  }

  return createIndex(parentItem->row(), 0, parentItem);
}

int TreeModel::rowCount(const QModelIndex &parent) const {
  TreeItem *parentItem;
  if (parent.column() > 0) {
    return 0;
  }

  if (!parent.isValid()) {
    parentItem = rootItem;
  } else {
    parentItem = static_cast<TreeItem *>(parent.internalPointer());
  }

  return parentItem->rowCount();
}

void TreeModel::appendRow(const QStringList &data) {
  auto variantData = stringListToVariantList(data);
  rootItem->appendChild(new TreeItem(variantData, rootItem));
}

void TreeModel::setHorizontalHeaderLabels(const QStringList &labels) {
  auto variantData = stringListToVariantList(labels);
  rootItem->setData(variantData);
}

TreeItem *TreeModel::itemFromIndex(const QModelIndex &index) const {
  if (index.isValid()) {
    auto item = static_cast<TreeItem *>(index.internalPointer());
    if (item) {
      return item;
    }
  }
  return rootItem;
}

TreeItem *TreeModel::item(int row, int column) const {
  return itemFromIndex(index(row, column));
}

QModelIndex TreeModel::indexFromItem(TreeItem *item) const {
  if (item == rootItem) {
    return QModelIndex();
  }

  return createIndex(item->row(), 0, item);
}

bool TreeModel::removeColumns(int position, int columns,
                              const QModelIndex &parent) {
  beginRemoveColumns(parent, position, position + columns - 1);
  try {
    rootItem->removeColumns(position, columns);
  } catch (std::out_of_range &e) {
    qDebug() << e.what();
    endRemoveColumns();
    return false;
  }
  endRemoveColumns();

  if (rootItem->columnCount() == 0) {
    removeRows(0, rowCount());
  }

  return true;
}

bool TreeModel::removeRows(int position, int rows, const QModelIndex &parent) {
  auto parentItem = itemFromIndex(parent);
  if (!parentItem) {
    return false;
  }

  beginRemoveRows(parent, position, position + rows - 1);
  try {
    parentItem->removeRows(position, rows);
  } catch (std::out_of_range &e) {
    qDebug() << e.what();
    endRemoveRows();
    return false;
  }
  endRemoveRows();

  return true;
}

bool TreeModel::setData(const QModelIndex &index, const QVariant &value,
                        int role) {
  if (role != Qt::EditRole) {
    return false;
  }

  auto item = itemFromIndex(index);
  try {
    item->setData(index.column(), value);
  } catch (std::out_of_range &e) {
    qDebug() << e.what();
    return false;
  }
  emit dataChanged(index, index, {Qt::DisplayRole, Qt::EditRole});
  return true;
}

bool TreeModel::setHeaderData(int section, Qt::Orientation orientation,
                              const QVariant &value, int role) {
  if (role != Qt::EditRole || orientation != Qt::Horizontal) {
    return false;
  }

  try {
    rootItem->setData(section, value);
  } catch (std::out_of_range &e) {
    qDebug() << e.what();
    return false;
  }

  emit headerDataChanged(orientation, section, section);
  return true;
}

bool TreeModel::insertColumns(int position, int columns,
                              const QModelIndex &parent) {
  beginInsertColumns(parent, position, position + columns - 1);
  try {
    rootItem->insertColumns(position, columns);
  } catch (std::out_of_range &e) {
    qDebug() << e.what();
    endInsertColumns();
    return false;
  }
  endInsertColumns();
  return true;
}

bool TreeModel::insertRows(int position, int rows, const QModelIndex &parent) {
  auto parentItem = itemFromIndex(parent);
  if (!parentItem) {
    return false;
  }

  beginInsertRows(parent, position, position + rows - 1);
  try {
    parentItem->insertRows(position, rows, rootItem->columnCount());
  } catch (std::out_of_range &e) {
    qDebug() << e.what();
    endInsertRows();
    return false;
  }
  endInsertRows();

  return true;
}

File: src/mvc/python/tree_model/tree_model.py
from PyQt6.QtCore import QModelIndex, QVariant, Qt, QAbstractItemModel, QObject


class TreeItem(QObject):
    def __init__(self, data, parent=None):
        super().__init__(parent)
        self._data = data
        self.parent_item = parent
        self.child_items = []

    def data(self, column=None):
        if column is None:
            return self._data
        if column < 0 or column >= self.column_count():
            return None
        return self._data[column]

    def setData(self, column, value):
        if column < 0 or column >= self.column_count():
            return
        self._data[column] = value

    def append_child(self, item):
        if not item:
            return
        current_parent = item.parent()
        if current_parent:
            current_parent.remove_child(item)
        item.setParent(self)
        self.child_items.append(item)

    def child(self, row):
        if row < 0 or row >= len(self.child_items):
            return None
        return self.child_items[row]

    def child_from_row(self, child):
        return self.child_items.indexOf(child)

    def children_items(self):
        return self.child_items

    def row_count(self):
        return len(self.child_items)

    def column_count(self):
        return len(self._data)

    def children_column_count(self):
        max_count = 0
        for child in self.child_items:
            max_count = max(max_count, child.column_count())
        return max_count

    def parent(self):
        return self.parent_item

    def row(self):
        if self.parent_item:
            return self.parent_item.child_items.index(self)
        return 0

    def remove_row(self, row):
        if row < 0 or row >= self.row_count():
            return
        self.child_items.remove(row)

    def remove_rows(self, row, count):
        for i in range(count):
            self.child_items.remove(row)

    def remove_child(self, child):
        if child in self.child_items:
            self.child_items.remove(child)

    def remove_all_children(self):
        self.child_items.clear()


class TreeModel(QAbstractItemModel):
    def __init__(self, headers, parent=None):
        super().__init__(parent)
        root_data = [header for header in headers]
        self.root_item = TreeItem(root_data)
        self.readOnly = False

    def setReadOnly(self, flag):
        self.readOnly = flag

    def isReadOnly(self):
        return self.readOnly

    def columnCount(self, parent):
        if parent.isValid():
            return parent.internalPointer().children_column_count()
        return self.root_item.column_count()

    def data(self, index, role):
        if not index.isValid():
            return QVariant()

        if role == Qt.ItemDataRole.DisplayRole or role == Qt.ItemDataRole.EditRole:
            item = self.itemFromIndex(index)
            return item.data(index.column())

        return QVariant()

    def flags(self, index):
        if not index.isValid():
            return Qt.ItemFlag.NoItemFlags
        if self.readOnly:
            return Qt.ItemFlag.ItemIsEnabled | Qt.ItemFlag.ItemIsSelectable
        if self.hasChildren(index):
            return Qt.ItemFlag.ItemIsEnabled | Qt.ItemFlag.ItemIsSelectable
        else:
            return (
                Qt.ItemFlag.ItemIsEnabled
                | Qt.ItemFlag.ItemIsSelectable
                | Qt.ItemFlag.ItemIsEditable
            )
        return Qt.ItemFlag.ItemIsEnabled

    def headerData(self, section, orientation, role):
        if (
            orientation == Qt.Orientation.Horizontal
            and role == Qt.ItemDataRole.DisplayRole
        ):
            return self.root_item.data(section)
        return QVariant()

    def index(self, row, column, parent):
        if not self.hasIndex(row, column, parent):
            return QModelIndex()

        parent_item = (
            self.root_item if not parent.isValid() else parent.internalPointer()
        )
        child_item = parent_item.child(row)

        if child_item:
            return self.createIndex(row, column, child_item)

        return QModelIndex()

    def parent(self, index):
        if not index.isValid():
            return QModelIndex()

        child_item = index.internalPointer()
        parent_item = child_item.parent()

        if parent_item == self.root_item:
            return QModelIndex()

        return self.createIndex(parent_item.row(), 0, parent_item)

    def rowCount(self, parent):
        if parent.column() > 0:
            return 0
        if not parent.isValid():
            parent_item = self.root_item
        else:
            parent_item = parent.internalPointer()
        return parent_item.row_count()

    def appendRow(self, data):
        self.root_item.append_child(TreeItem(data, self.root_item))

    def setHorizontalHeaderLabels(self, labels):
        self.root_item.setData(labels)

    def itemFromIndex(self, index):
        if index.isValid():
            item = index.internalPointer()
        if item:
            return item
        return self.root_item

    def item(self, row, column):
        return self.itemFromIndex(self.index(row, column))

    def indexFromItem(self, item):
        if item == self.root_item:
            return QModelIndex()
        return self.createIndex(item.row(), 0, item)

    def removeColumns(self, position, columns, parent):
        self.beginRemoveColumns(parent, position, position + columns - 1)
        try:
            self.root_item.removeColumns(position, columns)
        except Exception as e:
            self.endRemoveColumns()
            return False
        self.endRemoveColumns()

        if self.root_item.column_count() == 0:
            self.removeRows(0, self.rowCount())

        return True

    def removeRows(self, position, rows, parent):
        parent_item = self.itemFromIndex(parent)
        if not parent_item:
            return False
        self.beginRemoveRows(parent, position, position + rows - 1)
        try:
            parent_item.remove_rows(position, rows)

        except Exception as e:
            self.endRemoveRows()
            return False
        self.endRemoveRows()

        return True

    def setData(self, index, value, role):
        if role != Qt.ItemDataRole.EditRole:
            return False
        item = self.itemFromIndex(index)
        try:
            item.setData(index.column(), value)
        except Exception as e:
            return False
        self.dataChanged.emit(
            index, index, [Qt.ItemDataRole.DisplayRole, Qt.ItemDataRole.EditRole]
        )
        return True

    def setHeaderData(self, section, orientation, value, role):
        if role != Qt.ItemDataRole.EditRole or orientation != Qt.Orientation.Horizontal:
            return False
        try:
            self.root_item.setData(section, value)
        except Exception as e:
            return False
        self.headerDataChanged.emit(orientation, section, section)
        return True

    def insertColumns(self, position, columns, parent):
        self.beginInsertColumns(parent, position, position + columns - 1)
        try:
            self.root_item.insertColumns(position, columns)
        except Exception as e:
            self.endInsertColumns()
            return False
        self.endInsertColumns()
        return True

    def insertRows(self, position, rows, parent):
        parent_item = self.itemFromIndex(parent)
        if not parent_item:
            return False
        self.beginInsertRows(parent, position, position + rows - 1)
        try:
            parent_item.insertRows(position, rows, self.root_item.column_count())
        except Exception as e:
            self.endInsertRows()
            return False
        self.endInsertRows()
        return True
